ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferReadableStreamChannel.java

```diff
@@ -65,6 +74,30 @@ public long getSize() {
   }
 
   @Override
+  public void setDigestAlgorithm(String digestAlgorithm)
+      throws NoSuchAlgorithmException {
+    if (digest != null && digest.getAlgorithm().equals(digestAlgorithm)) {
+      return;
+    }
+    digest = MessageDigest.getInstance(digestAlgorithm);
+    digestBytes = null;
+  }
+
+  @Override
+  public byte[] getDigest() {
+    if (digest == null) {
+      return null;
+    } else if (digestBytes == null) {
+      int savedPosition = buffer.position();
+      buffer.position(startPos);
+      digest.update(buffer);
+      buffer.position(savedPosition);
+      digestBytes = digest.digest();
```

Comment:
I think i'll retain this for now because it doesn't create a new object and the implementation is not very complex.

Orthogonal to that, I'm hesitant about using `mark()` in cases where an external API is called. In this particular case, there are ways to work around it, but just wanted to highlight my thoughts about this. The external API might also `mark()` and `reset()` internally to achieve some purpose. In that case, wouldn't the code you suggest turn out to be incorrect? This is unrelated to your `slice()` suggestion which is sound.